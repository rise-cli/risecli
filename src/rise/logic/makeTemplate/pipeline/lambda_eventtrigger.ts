export function buildEventInput(eventInput: any) {
    let v = '{'
    Object.keys(eventInput).forEach((k: string, i) => {
        v = v + `${k}: props.${eventInput[k]}`
        if (Object.keys(eventInput).length > i + 1) {
            v = v + ','
        }
    })

    v = v + '}'
    return v
}

export function makeLambdaEventTrigger({
    key,
    apiName,
    eventName,
    index,
    query,
    eventInput
}: {
    key: string
    apiName: string
    eventName: string
    index: number
    query: string
    eventInput: any
}) {
    let value = buildEventInput(eventInput)

    const code = `
            const env = require("process").env;
            const AWS = require("aws-sdk");
            const URL = require("url");
            const https = require('https');

            AWS.config.update({
            region: process.env.REGION,
            credentials: new AWS.Credentials(
                env.AWS_ACCESS_KEY_ID,
                env.AWS_SECRET_ACCESS_KEY,
                env.AWS_SESSION_TOKEN
            ),
            });

            module.exports.handler = (props) => {
                const body = {
                    query: \`${query}\`,
                    variables: {
                        input: ${value}
                    }
                }
                const uri = URL.parse(process.env.ENDPOINT);
                const httpRequest = new AWS.HttpRequest(uri.href, process.env.REGION);
                httpRequest.headers.host = uri.host;
                httpRequest.headers["Content-Type"] = "application/json";
                httpRequest.method = "POST";
                httpRequest.body = JSON.stringify(body);

                const signer = new AWS.Signers.V4(httpRequest, "appsync", true);
                signer.addAuthorization(AWS.config.credentials, AWS.util.date.getDate());

                const options = {
                    hostname: uri.href.slice(8, uri.href.length - 8),
                    path: '/graphql',
                    method: httpRequest.method,
                    body: httpRequest.body,
                    headers: httpRequest.headers,
                };

                const req = https.request(options, res => {

                res.on('data', d => {
                        process.stdout.write(d)
                    })
                })

                req.on('error', error => {
                    console.error(error.message)
                })

                req.write(JSON.stringify(body))
                req.end()
            }`

    return {
        Resources: {
            [`TriggerSubscriptionFunction${apiName}${key}${eventName}${index}`]:
                {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Runtime: 'nodejs12.x',
                        Handler: 'index.handler',
                        Role: {
                            'Fn::GetAtt': [
                                `TriggerSubscriptionRole${apiName}${key}${eventName}${index}`,
                                'Arn'
                            ]
                        },
                        Code: {
                            ZipFile: code
                        },
                        Environment: {
                            Variables: {
                                REGION: {
                                    'Fn::Sub': ['${AWS::Region}', {}]
                                },
                                ENDPOINT: {
                                    'Fn::GetAtt': ['GraphQlApi', 'GraphQLUrl']
                                }
                            }
                        }
                    }
                },
            [`TriggerSubscriptionRole${apiName}${key}${eventName}${index}`]: {
                Type: 'AWS::IAM::Role',
                Properties: {
                    RoleName: `TriggerSubscriptionRole${apiName}${key}${eventName}${index}`,
                    AssumeRolePolicyDocument: {
                        Version: '2012-10-17',
                        Statement: [
                            {
                                Effect: 'Allow',
                                Action: ['sts:AssumeRole'],
                                Principal: {
                                    Service: ['lambda.amazonaws.com']
                                }
                            }
                        ]
                    },

                    Policies: [
                        {
                            PolicyName: `TriggerSubscriptionPolicy${apiName}`,
                            PolicyDocument: {
                                Version: '2012-10-17',

                                Statement: [
                                    {
                                        Effect: 'Allow',
                                        Action: ['appsync:GraphQL'],
                                        Resource: [
                                            {
                                                'Fn::Join': [
                                                    '',
                                                    [
                                                        {
                                                            'Fn::GetAtt': [
                                                                'GraphQlApi',
                                                                'Arn'
                                                            ]
                                                        },
                                                        '/types/*'
                                                    ]
                                                ]
                                            }
                                        ]
                                    }
                                ]
                            }
                        }
                    ]
                }
            }
        },
        Outputs: {}
    }
}
