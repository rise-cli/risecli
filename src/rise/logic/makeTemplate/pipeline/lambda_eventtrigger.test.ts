import { buildEventInput, makeLambdaEventTrigger } from './lambda_eventtrigger'

test('buildEventInput will create valid js object', () => {
    const result = buildEventInput({
        one: 'one',
        two: 'two',
        three: 'three'
    })

    expect(result).toBe('{one: props.one,two: props.two,three: props.three}')
})

test('makeLambdaEventTrigger will render correctly', () => {
    const result = makeLambdaEventTrigger({
        apiName: 'apiName',
        eventName: 'eventName',
        index: 2,
        query: `query`,
        eventInput: {
            one: 'one',
            two: 'two',
            three: 'three'
        }
    })

    expect(result).toMatchSnapshot()
})
