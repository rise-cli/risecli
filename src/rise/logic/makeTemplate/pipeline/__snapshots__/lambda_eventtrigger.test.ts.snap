// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeLambdaEventTrigger will render correctly 1`] = `
Object {
  "Outputs": Object {},
  "Resources": Object {
    "TriggerSubscriptionFunctionapiNameeventName2": Object {
      "Properties": Object {
        "Code": Object {
          "ZipFile": "
            const env = require(\\"process\\").env;
            const AWS = require(\\"aws-sdk\\");
            const URL = require(\\"url\\");
            const https = require('https');

            AWS.config.update({
            region: process.env.REGION,
            credentials: new AWS.Credentials(
                env.AWS_ACCESS_KEY_ID,
                env.AWS_SECRET_ACCESS_KEY,
                env.AWS_SESSION_TOKEN
            ),
            });

            module.exports.handler = (props) => {
                const body = {
                    query: \`query\`,
                    variables: {
                        input: {one: props.one,two: props.two,three: props.three}
                    }
                }
                const uri = URL.parse(process.env.ENDPOINT);
                const httpRequest = new AWS.HttpRequest(uri.href, process.env.REGION);
                httpRequest.headers.host = uri.host;
                httpRequest.headers[\\"Content-Type\\"] = \\"application/json\\";
                httpRequest.method = \\"POST\\";
                httpRequest.body = JSON.stringify(body);

                const signer = new AWS.Signers.V4(httpRequest, \\"appsync\\", true);
                signer.addAuthorization(AWS.config.credentials, AWS.util.date.getDate());

                const options = {
                    hostname: uri.href.slice(8, uri.href.length - 8),
                    path: '/graphql',
                    method: httpRequest.method,
                    body: httpRequest.body,
                    headers: httpRequest.headers,
                };

                const req = https.request(options, res => {

                res.on('data', d => {
                        process.stdout.write(d)
                    })
                })

                req.on('error', error => {
                    console.error(error.message)
                })

                req.write(JSON.stringify(body))
                req.end()
            }",
        },
        "Environment": Object {
          "Variables": Object {
            "ENDPOINT": Object {
              "Fn::GetAtt": Array [
                "GraphQlApi",
                "GraphQLUrl",
              ],
            },
            "REGION": Object {
              "Fn::Sub": Array [
                "\${AWS::Region}",
                Object {},
              ],
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TriggerSubscriptionRoleapiNameeventName2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TriggerSubscriptionRoleapiNameeventName2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "appsync:GraphQL",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          Object {
                            "Fn::GetAtt": Array [
                              "GraphQlApi",
                              "Arn",
                            ],
                          },
                          "/types/*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "TriggerSubscriptionPolicyapiName",
          },
        ],
        "RoleName": "TriggerSubscriptionRoleapiNameeventName2",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
